### PR TITLE
filters - lookup for value_from once instead of per resource in case where jmespath returns None

### PR DESCRIPTION
**Pull request by thisisshi**
When executing a policy with a value from filter, if the value returned by a jmespath search is null/None, the policy will attempt to retrieve the contents from the file for each resource. This can result in inconsistent behavior: 
- if the value of the jmespath search is None and later becomes a value during the policy execution, there is inconsistency on how the resources in an account are evaluated. In this case, `self.v` should be static during the evaluation of the filter.
- Results in long executions if the value of a lookup is supposed to be "null" as each lookup will open the file (remotely in s3 or on disk) and do the jmespath search.
- None is a often-used value when doing things like tag-compliance policies, in the current case this would also save a few additional calls as `self.v = self.data.get('value')` would only be called once per value filter instead of once per resource.

This PR introduces a content_semaphore attr in the Filter object to prevent multiple lookups of a value during runtime and logging when the jmespath search returns None as it is not a typical use-case and is likely a result of an error on the path or in the file used in the `value_from` filter.

### CODE REVIEW

**Comment by kapilt** [comment]
so two thoughts, the remote file should be cached, the deserialize into memory per resource isn't ideal though, a cache here probably belongs on the value_from implementation rather then extending value_filter with additional cases.

**Comment by thisisshi** [comment]
The policy pulls down a 1.1MB json file from S3 in the value_from filter, Benchmarks:

PR:

```
$ custodian run test.yml -s output --region us-east-1 -v --dryrun --debug --cache-period 0
2018-06-28 10:17:32,825: custodian.cache:DEBUG Disabling cache
2018-06-28 10:17:32,825: custodian.commands:DEBUG Loaded file test.yml. Contains 1 policies
2018-06-28 10:17:33,570: custodian.cache:DEBUG Disabling cache
2018-06-28 10:17:33,571: custodian.output:DEBUG Storing output with <DirectoryOutput to dir:output/valuefrom-test>
2018-06-28 10:17:33,571: custodian.policy:DEBUG Running policy valuefrom-test resource: ebs-snapshot region:us-east-1 c7n:0.8.28.2
2018-06-28 10:17:39,127: custodian.resolver:WARNING ValueFilter: accounts."xxxxxxxxxxxx".ebs.snapshots.*[][] key returned None
2018-06-28 10:17:39,134: custodian.resources.snapshot:DEBUG Filtered from 668 to 0 snapshot
2018-06-28 10:17:39,137: custodian.policy:INFO policy: valuefrom-test resource:ebs-snapshot region:us-east-1 count:0 time:5.57
2018-06-28 10:17:39,137: custodian.output:DEBUG metric:ResourceCount Count:0 policy:valuefrom-test restype:ebs-snapshot scope:policy
2018-06-28 10:17:39,138: custodian.cache:DEBUG Disabling cache
```
5.57 seconds for execution

Master:
```
$ custodian run test.yml -s output --region us-east-1 -v --dryrun --debug --cache-period 0
2018-06-28 10:17:53,718: custodian.cache:DEBUG Disabling cache
2018-06-28 10:17:53,719: custodian.commands:DEBUG Loaded file test.yml. Contains 1 policies
2018-06-28 10:17:54,472: custodian.cache:DEBUG Disabling cache
2018-06-28 10:17:54,472: custodian.output:DEBUG Storing output with <DirectoryOutput to dir:output/valuefrom-test>
2018-06-28 10:17:54,472: custodian.policy:DEBUG Running policy valuefrom-test resource: ebs-snapshot region:us-east-1 c7n:0.8.28.2
2018-06-28 10:32:29,641: custodian.resources.snapshot:DEBUG Filtered from 668 to 0 snapshot
2018-06-28 10:32:29,643: custodian.policy:INFO policy: valuefrom-test resource:ebs-snapshot region:us-east-1 count:0 time:875.17
2018-06-28 10:32:29,643: custodian.output:DEBUG metric:ResourceCount Count:0 policy:valuefrom-test restype:ebs-snapshot scope:policy
2018-06-28 10:32:29,644: custodian.cache:DEBUG Disabling cache
```
14.5 minutes for execution

File: c7n/resolver.py [hunk 1]
```diff
@@ -134,7 +137,10 @@ def get_values(self):
         if format == 'json':
             data = json.loads(contents)
             if 'expr' in self.data:
-                return jmespath.search(self.data['expr'], data)
+                res = jmespath.search(self.data['expr'], data)
+                if res is None:
+                    log.warning('ValueFilter: %s key returned None' % self.data['expr'])
```
Comment (kapilt) [inline c7n/resolver.py]: s/ValueFilter/ValueFrom .. we reuse the resolver implementation across many filters/etc.

File: c7n/resolver.py [hunk 2]
```diff
@@ -144,7 +150,10 @@ def get_values(self):
                     return [d[self.data['expr']] for d in data]
                 data = list(data)
             if 'expr' in self.data:
-                return jmespath.search(self.data['expr'], data)
+                res = jmespath.search(self.data['expr'], data)
+                if res is None:
+                    log.warning('ValueFilter: %s key returned None' % self.data['expr'])
```
Comment (kapilt) [inline c7n/resolver.py]: ditto re ValueFrom

**Comment by kapilt** [comment]
thanks for detailed writeup and benchmarks, lgtm.

**Pull request merged**